// @injectivelabs/core-proto-ts/esm/injective/peggy/v1/proposal.ts
/* eslint-disable */
import _m0 from "protobufjs/minimal";

export interface BlacklistEthereumAddressesProposal {
  title: string;
  description: string;
  blacklistAddresses: string[];
}

export interface RevokeEthereumBlacklistProposal {
  title: string;
  description: string;
  blacklistAddresses: string[];
}

function createBaseBlacklistEthereumAddressesProposal(): BlacklistEthereumAddressesProposal {
  return { title: "", description: "", blacklistAddresses: [] };
}

export const BlacklistEthereumAddressesProposal = {
  encode(message: BlacklistEthereumAddressesProposal, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.title !== "") {
      writer.uint32(10).string(message.title);
    }
    if (message.description !== "") {
      writer.uint32(18).string(message.description);
    }
    for (const v of message.blacklistAddresses) {
      writer.uint32(26).string(v!);
    }
    return writer;
  },

  decode(input: _m0.Reader | Uint8Array, length?: number): BlacklistEthereumAddressesProposal {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseBlacklistEthereumAddressesProposal();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.title = reader.string();
          break;
        case 2:
          message.description = reader.string();
          break;
        case 3:
          message.blacklistAddresses.push(reader.string());
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },

  fromJSON(object: any): BlacklistEthereumAddressesProposal {
    return {
      title: isSet(object.title) ? String(object.title) : "",
      description: isSet(object.description) ? String(object.description) : "",
      blacklistAddresses: Array.isArray(object?.blacklistAddresses)
        ? object.blacklistAddresses.map((e: any) => String(e))
        : [],
    };
  },

  toJSON(message: BlacklistEthereumAddressesProposal): unknown {
    const obj: any = {};
    message.title !== undefined && (obj.title = message.title);
    message.description !== undefined && (obj.description = message.description);
    if (message.blacklistAddresses) {
      obj.blacklistAddresses = message.blacklistAddresses.map((e) => e);
    } else {
      obj.blacklistAddresses = [];
    }
    return obj;
  },

  create(base?: DeepPartial<BlacklistEthereumAddressesProposal>): BlacklistEthereumAddressesProposal {
    return BlacklistEthereumAddressesProposal.fromPartial(base ?? {});
  },

  fromPartial(object: DeepPartial<BlacklistEthereumAddressesProposal>): BlacklistEthereumAddressesProposal {
    const message = createBaseBlacklistEthereumAddressesProposal();
    message.title = object.title ?? "";
    message.description = object.description ?? "";
    message.blacklistAddresses = object.blacklistAddresses?.map((e) => e) || [];
    return message;
  },
};

function createBaseRevokeEthereumBlacklistProposal(): RevokeEthereumBlacklistProposal {
  return { title: "", description: "", blacklistAddresses: [] };
}

export const RevokeEthereumBlacklistProposal = {
  encode(message: RevokeEthereumBlacklistProposal, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.title !== "") {
      writer.uint32(10).string(message.title);
    }
    if (message.description !== "") {
      writer.uint32(18).string(message.description);
    }
    for (const v of message.blacklistAddresses) {
      writer.uint32(26).string(v!);
    }
    return writer;
  },

  decode(input: _m0.Reader | Uint8Array, length?: number): RevokeEthereumBlacklistProposal {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseRevokeEthereumBlacklistProposal();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.title = reader.string();
          break;
        case 2:
          message.description = reader.string();
          break;
        case 3:
          message.blacklistAddresses.push(reader.string());
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },

  fromJSON(object: any): RevokeEthereumBlacklistProposal {
    return {
      title: isSet(object.title) ? String(object.title) : "",
      description: isSet(object.description) ? String(object.description) : "",
      blacklistAddresses: Array.isArray(object?.blacklistAddresses)
        ? object.blacklistAddresses.map((e: any) => String(e))
        : [],
    };
  },

  toJSON(message: RevokeEthereumBlacklistProposal): unknown {
    const obj: any = {};
    message.title !== undefined && (obj.title = message.title);
    message.description !== undefined && (obj.description = message.description);
    if (message.blacklistAddresses) {
      obj.blacklistAddresses = message.blacklistAddresses.map((e) => e);
    } else {
      obj.blacklistAddresses = [];
    }
    return obj;
  },

  create(base?: DeepPartial<RevokeEthereumBlacklistProposal>): RevokeEthereumBlacklistProposal {
    return RevokeEthereumBlacklistProposal.fromPartial(base ?? {});
  },

  fromPartial(object: DeepPartial<RevokeEthereumBlacklistProposal>): RevokeEthereumBlacklistProposal {
    const message = createBaseRevokeEthereumBlacklistProposal();
    message.title = object.title ?? "";
    message.description = object.description ?? "";
    message.blacklistAddresses = object.blacklistAddresses?.map((e) => e) || [];
    return message;
  },
};

type Builtin = Date | Function | Uint8Array | string | number | boolean | undefined;

type DeepPartial<T> = T extends Builtin ? T
  : T extends Array<infer U> ? Array<DeepPartial<U>> : T extends ReadonlyArray<infer U> ? ReadonlyArray<DeepPartial<U>>
  : T extends {} ? { [K in keyof T]?: DeepPartial<T[K]> }
  : Partial<T>;

function isSet(value: any): boolean {
  return value !== null && value !== undefined;
}
